### ISSUE TITLE
Support non-nullable value types with the InitRequiredForNotNullAttribute

### ISSUE THREAD

**Issue by jmevel** [cezarypiatek/CSharpExtensions#45]
The `InitRequiredForNotNullAttribute` is really great, we've activated it in all our projects but there is one missing piece in my opinion: the non-nullable value types!

Consider this code:

```csharp
using System;

class Program
{
    static void Main()
    {
        _ = new Person
        {
            Name = "Foo"
        };
    }
}

class Person
{
    public string Name { get; set; }

    public DateTime BirthDate { get; set; }
}
```

I'm expecting to get the CSE001 error but actually not because `BirthDate` is a value type.

Would that be possible to implement the same behavior as for non-nullable reference types?
That would be awesome.

Thank you

**Comment by cezarypiatek** [comment]
I will take a look at that later, thanks for reporting this improvement

**Comment by cezarypiatek** [comment]
Not sure why I missed that in the first place. Should be fixed in v4.0.0
I will be grateful if you could test it and let me know if everything is working according to your expectations.

**Comment by vsilvar** [comment]
@cezarypiatek 
Tested this change and it's working, but I found something I wouldn't expect.

Given the following code:
```c#
using SmartAnalyzers.CSharpExtensions.Annotations;

[assembly: InitRequiredForNotNull]
namespace AnalyzerTest
{
    class Program
    {
        static void Main()
        {
            _ = new Person
            {
                Name = "Foo",
                Age = 18
            };
        }
    }

    class Person
    {
        public string Name { get; set; }
        public int Age { get; set; }
        public bool IsAdmin { get; set; } = false;
    }
}
```

I get `CSE001 - Missing initialization for properties: - IsAdmin`, which I wouldn't expect as a default value is being set.
At the very least an `[InitOptional]` attribute could exist to allow disabling the initialization checks for a single property.

**Comment by jmevel** [comment]
Cool, I updated to version 4 and it's now showing the error when we forget to initialize some non-nullable value types.

But @vsilvar has a very good point, I didn't think about properties that are already initialized like this. 
The `InitOptional` attribute seems like a good idea to solve this specific case.

**Comment by cezarypiatek** [comment]
I've just added `[InitOptional]` which allows for excluding a specific member or whole type from the mandatory init imposed on the higher level by [InitRequired] or [InitRequiredForNotNullAttribute]. Should be available in v4.1

Please let me know how do you find the current behavior.
